### PR TITLE
[feature] support npu allocator

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

New features
### PR changes

Others
### Describe

Support npu allocator for Ascend Card

Usage
```
ptr = tensor.mutable_data<float>(framework::make_ddim({1, 2, 3}), platform::NPUPlace(0));
```

- Currently, only naive_best_fit_allocator is supported.
- Reuse the gpu flags to control memory allocation.
  	- FLAGS_selected_gpus
  	- FLAGS_reallocate_gpu_memory_in_mb
	- FLAGS_initial_gpu_memory_in_mb
	- FLAGS_fraction_of_gpu_memory_to_use
- Add `-DWITH_ASCEND_CL=ON` to cmake 

### Main Related files
- Basic and common NPU-related structures and functions
  	- npu_info.cc/h，add some NPU-related common functions
  	- device_guard.cc/h, add NPUDeviceContext
  	- enforce.h, add PADDLE_ENFORCE_NPU_SUCCESS
  	- place.cc/h, add NPUPlace
  	- init.cc, initilize NPU Device
  	- pybind.cc, add `core. _npu_finalize` for NPU Device to do finalize exlpicitly
  	-  flags.cc, make some GPU-related flags visible for NPU
- NPU Allocator 
	- system_allocator.cc/h, add NPU allocator for NaiveBestFitAllocator
	- buddy_allocator.cc/h, update to support NPU allocator
	- naive_best_fit_allocator.cc/h, update to support NPU allocator
	- allocator_facede.cc, allocator interface
	- monitor.cc/h, supoprt NPU monitor for memory allocation
	- npu_allocator.cc/h, add NPU allocator for AutoGrowthAllocator (for future use)
- Unittest
	- system_allocator_test.cc
	- buddy_allocator_test.cc
	- naive_best_fit_allocator_test.cc
	- tensor_test.cc
- Some cmake files
- Some template partial specialization of place, add specialization for NPUPlace (otherwise failed to compile)
	- dlpack_tensor.cc, 
	- tensor_util.cc
	- gradient_accumulator.cc
	- math_function.cc

### CODE REVIEW

File: paddle/fluid/memory/allocation/allocator_facade.cc [hunk 1]
```diff
@@ -42,6 +42,7 @@
 #ifdef PADDLE_WITH_XPU
 #include "paddle/fluid/platform/xpu_info.h"
 #endif
+#include "paddle/fluid/platform/npu_info.h"
```
Comment (phlrain) [inline paddle/fluid/memory/allocation/allocator_facade.cc:45]: #ifdef PADDLE_WITH_ASCEND_CL

File: paddle/fluid/memory/detail/system_allocator.cc [hunk 2]
```diff
@@ -35,6 +35,7 @@ limitations under the License. */
 #include "paddle/fluid/platform/cpu_info.h"
 #include "paddle/fluid/platform/enforce.h"
 #include "paddle/fluid/platform/gpu_info.h"
+#include "paddle/fluid/platform/npu_info.h"
```
Comment (phlrain) [inline paddle/fluid/memory/detail/system_allocator.cc:38]: #ifdef PADDLE_WITH_ASCEND_CL

File: paddle/fluid/memory/allocation/allocator_facade.cc (hunk 1 above)
Comment (zhiqiu) [reply paddle/fluid/memory/allocation/allocator_facade.cc:45]: I will refine this kind includes in the future, for this line, `paddle/fluid/platform/npu_info.h` is empty if not compiled with ASCEND_CL, so it is ok.

File: paddle/fluid/memory/detail/system_allocator.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/memory/detail/system_allocator.cc:38]: same above.

**Pull request merged**